### PR TITLE
ENH JTFS tensorflow frontend

### PR DESCRIPTION
**Pull request by lylyhan**
tensorflow frontend:
completed the following files:
- tensorflow_frontend.py: added JTFS frontend class
- test_tensorflow_scattering1d.py: added differentiability test for scattering 1d, as in test_torch_scattering1d.py
- test_timefrequency_scattering_tensorflow.py: added test_jtfs_build, test_Q, test_check_runtime_args, test_jtfs_create_filters, test_time_scattering_widthfirst, test_frequency_scattering, test_joint_timefrequency_scattering, test_differentiability_jtfs, test_jtfs_numpy. As in test_timefrequency_scattering.py for pytorch.

### CODE REVIEW

File: kymatio/scattering1d/frontend/tensorflow_frontend.py [hunk 1]
```diff
@@ -15,7 +15,27 @@ def __init__(self, J, shape, Q=1, T=None, max_order=2, oversampling=0,
...
 
+
 ScatteringTensorFlow1D._document()
 
 
-__all__ = ['ScatteringTensorFlow1D']
+class TimeFrequencyScatteringTensorflow(
+        ScatteringTensorFlow, TimeFrequencyScatteringBase):
+    def __init__(self, *, J, J_fr, shape, Q, T=None, oversampling=0, Q_fr=1,
+            F=None, oversampling_fr=0, out_type='array', backend='tensorflow', name="TimeFrequencyScatteringScattering"):
```
Comment (lostanlen) [inline kymatio/scattering1d/frontend/tensorflow_frontend.py:25]: try unifying between single quotes `out_type='array'` and double quotes `name="TimeFrequencyScattering"`

also ... `name="TimeFrequencyScattering"`, not `"TimeFrequencyScatteringScattering"`

File: kymatio/scattering1d/frontend/tensorflow_frontend.py [hunk 2]
```diff
@@ -15,7 +15,27 @@ def __init__(self, J, shape, Q=1, T=None, max_order=2, oversampling=0,
...
+            oversampling_fr=oversampling_fr, out_type=out_type, backend=backend)
+        TimeFrequencyScatteringBase._instantiate_backend(
+            self, 'kymatio.scattering1d.backend.')
+        TimeFrequencyScatteringBase.build(self)
+        TimeFrequencyScatteringBase.create_filters(self)
+    
+
+
+
+__all__ = ['ScatteringTensorflow1D', 'TimeFrequencyScatteringTensorflow']
```
Comment (lostanlen) [inline kymatio/scattering1d/frontend/tensorflow_frontend.py:38]: capital F in TensorFlow SVP

weird that your tests don't fail??

File: tests/scattering1d/test_tensorflow_scattering1d.py [hunk 3]
```diff
@@ -6,11 +6,21 @@
 import tensorflow as tf
 
 
+
 backends = []
+gpu_available = tf.test.is_gpu_available()
+
+if not gpu_available:
+    Warning('GPU not available for tensorflow.')
```
Comment (lostanlen) [inline tests/scattering1d/test_tensorflow_scattering1d.py:14]: TensorFlow

File: kymatio/scattering1d/frontend/tensorflow_frontend.py (hunk 2 above)
Comment (lylyhan) [reply kymatio/scattering1d/frontend/tensorflow_frontend.py]: i have incorporated your feedback and pushed again. Indeed the tests didn't fail for some reasons..

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+    assert "Q_fr must be an integer or 1-tuple" in ve.value.args[0]
+
+    with pytest.raises(NotImplementedError) as ve:
+        TimeFrequencyScatteringBase(Q_fr=(1, 2), **jtfs_kwargs).build()
+    assert "Q_fr must be an integer or 1-tuple" in ve.value.args[0]
+
+
+def test_check_runtime_args():
+    kwargs = dict(J=10, J_fr=3, shape=4096, Q=8)
+    S = TimeFrequencyScatteringBase(**kwargs)
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:74]: why are we testing this here? i don't see tensorflow frontend and this is duplicate code?

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+    assert "integer" in ve.value.args[0]
+
+    with pytest.raises(ValueError) as ve:
+        S = TimeFrequencyScatteringBase(format="doesnotexist", **kwargs)
+        S.build()
+        S._check_runtime_args()
+    assert "format must be" in ve.value.args[0]
+
+
+def test_jtfs_create_filters():
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:127]: Again, i don't think it's necessary to test this here. It's backend agnostic.

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+    )
+    assert np.allclose(S1_width.numpy(), S1_depth.numpy())
+
+
+def test_frequency_scattering():
+    # Create S1 array as input
+    J = 5
+    shape = (4096,)
+    Q = 8
+    S = kymatio.Scattering1D(J=J, Q=Q, shape=shape, backend="numpy")
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:195]: why test `numpy` in this file?

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+    average_local = True
+    S_gen = scattering1d(U_0, S.backend, filters, S.oversampling, average_local)
+    S_1_dict = {path["n"]: path["coef"] for path in S_gen if len(path["n"]) == 1}
+    S_1 = S.backend.concatenate([S_1_dict[key] for key in sorted(S_1_dict.keys())])
+    X = {"coef": S_1, "n1_max": len(S_1_dict), "n": (-1,), "j": (-1,)}
+
+    # Define scattering object
+    J_fr = 3
+    jtfs = TimeFrequencyScatteringBase(
+        J=J, J_fr=J_fr, shape=shape, Q=Q, backend="numpy"
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:212]: why test numpy here?

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+
+
+def test_joint_timefrequency_scattering():
+    # Define scattering object
+    J = 8
+    J_fr = 3
+    shape = (4096,)
+    Q = 8
+    S = TimeFrequencyScatteringBase(
+        J=J, J_fr=J_fr, shape=shape, Q=Q, T=0, F=0, backend="numpy"
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:253]: why numpy?

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+        jtfs.oversampling_fr,
+        jtfs.average_fr == "local",
+        spinned=True,
+    )
+    for Y_fr in freq_gen:
+        assert Y_fr["coef"].shape[-1] == X["coef"].shape[-1]
+        assert Y_fr["n"] == (4,) + Y_fr["n_fr"]
+
+
+def test_joint_timefrequency_scattering():
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:246]: a lot of this code is duplicated for torch frontend. why don't we refactor it?

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+            loss = tf.norm(S["coef"])
+            assert tf.abs(loss) >= 0.0
+            losses.append(loss)
+    for loss in losses:
+        grad = tape.gradient(loss,x)
+        assert tf.reduce_max(tf.abs(grad)) > 0.0
+
+
+
+def test_jtfs_numpy():
```
[the remark this answers is not in the document]
Comment (cyrusvahidi) [reply tests/scattering1d/test_timefrequency_scattering_tensorflow.py:442]: also, it looks like #963 handles this

**Review by cyrusvahidi — CHANGES_REQUESTED** [review]
Differentiability looks good. I think we should address the code duplication with some refactoring. Also, this commit should not include numpy test, nor numpy tests within tensorflow test files.

File: kymatio/scattering1d/frontend/tensorflow_frontend.py (hunk 2 above)
Comment (lostanlen) [reply kymatio/scattering1d/frontend/tensorflow_frontend.py]: after cyrus's feedback, i think it's clear that your tests weren't testing the right thing

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+    # test different cases for Q_fr
+    with pytest.raises(ValueError) as ve:
+        TimeFrequencyScatteringBase(Q_fr=[1], **jtfs_kwargs).build()
+    assert "Q_fr must be an integer or 1-tuple" in ve.value.args[0]
+
+    with pytest.raises(NotImplementedError) as ve:
+        TimeFrequencyScatteringBase(Q_fr=(1, 2), **jtfs_kwargs).build()
+    assert "Q_fr must be an integer or 1-tuple" in ve.value.args[0]
+
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:70]: two lines between functions defs please

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    # Check order 1
+    keep_order1 = lambda item: (len(item[0]) == 1)
+    S1_width = dict(filter(keep_order1, W.items()))
+    assert len(S1_width) == 1
+    S1_width = S1_width[(-1,)]
+    S1_depth = S.backend.concatenate(
+        [S1_depth[key] for key in sorted(S1_depth.keys()) if len(key) == 1]
+    )
+    assert np.allclose(S1_width.numpy(), S1_depth.numpy())
+
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:134]: two lines between function defs

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    filters = [S.phi_f, S.psi1_f, S.psi2_f]
+
+    with tf.GradientTape(persistent=True) as tape:
+
+        x_shape = backend.shape(x)
+        batch_shape, signal_shape = x_shape[:-1], x_shape[-1:]
+        x_reshaped = backend.reshape_input(x, signal_shape)
+        U_0_in = backend.pad(x_reshaped, pad_left=S.pad_left, pad_right=S.pad_right)
+
+        jtfs_gen = joint_timefrequency_scattering(
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:161]: please format this according to some standard

File: tests/scattering1d/test_timefrequency_scattering_tensorflow.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        S.build()
+        S.create_filters()
+        x = tf.Variable(tf.random.normal(shape))
+    backend = S.backend
+    tf.random.set_seed(random_state) 
+
+    filters = [S.phi_f, S.psi1_f, S.psi2_f]
+
+    with tf.GradientTape(persistent=True) as tape:
+
```
Comment (cyrusvahidi) [inline tests/scattering1d/test_timefrequency_scattering_tensorflow.py:155]: extraneous line

**Review by cyrusvahidi — CHANGES_REQUESTED** [review]
please address the formatting

**Pull request merged**